In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
! head -4 /content/heart.csv

age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
52,1,0,125,212,0,1,168,0,1,2,2,3,0
53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
70,1,0,145,174,0,1,125,1,2.6,0,0,3,0


In [ ]:
! tail -10 /content/heart.csv

58,1,0,128,216,0,0,131,1,2.2,1,3,3,0
65,1,3,138,282,1,0,174,0,1.4,1,1,2,0
53,1,0,123,282,0,1,95,1,2,1,2,3,0
41,1,0,110,172,0,0,158,0,0,2,0,3,0
47,1,0,112,204,0,1,143,0,0.1,2,0,2,1
59,1,1,140,221,0,1,164,1,0,2,0,2,1
60,1,0,125,258,0,0,141,1,2.8,1,1,3,0
47,1,0,110,275,0,0,118,1,1,1,1,2,0
50,0,0,110,254,0,0,159,0,0,2,0,2,1
54,1,0,120,188,0,1,113,0,1.4,1,1,3,0


In [ ]:
df = pd.read_csv('/content/heart.csv')
df

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1020,59,1,1,140,221,0,1,164,1,0.0,2,0,2,1
1021,60,1,0,125,258,0,0,141,1,2.8,1,1,3,0
1022,47,1,0,110,275,0,0,118,1,1.0,1,1,2,0
1023,50,0,0,110,254,0,0,159,0,0.0,2,0,2,1


In [ ]:
df.shape # the shape function print the row * columns of dataset

(1025, 14)

In [ ]:
df.head() # prints the dfs first 10 line by default

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0


In [ ]:
df.info() # prints the metadata related to dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


In [ ]:
# let find total duplicate rows in dataset
duplicate_row_count = df.duplicated().sum()
duplicate_row_count

np.int64(723)

In [ ]:
# drop the exact duplicates in dataset (excluding first occurance)
df_clean = df.drop_duplicates()
df_clean

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
723,68,0,2,120,211,0,0,115,0,1.5,1,0,2,1
733,44,0,2,108,141,0,1,175,0,0.6,1,0,2,1
739,52,1,0,128,255,0,1,161,1,0.0,2,1,3,0
843,59,1,3,160,273,0,0,125,0,0.0,2,0,2,0


In [ ]:
print("cleaned df", df_clean.shape)
print("original df", df.shape)

cleaned df (302, 14)
original df (1025, 14)


In [ ]:
df.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

# columns in dataset
| **[Column](ca://s?q=Explain_UCI_Heart_Disease_Dataset_Columns)** | **Meaning** | **Clinical Importance** |
| --- | --- | --- |
| **[age](ca://s?q=Explain_age_in_heart_disease_dataset)** | Patient’s age (years) | Risk of heart disease increases with age. |
| **[sex](ca://s?q=Explain_sex_in_heart_disease_dataset)** | Male (1) or Female (0) | Men generally have higher risk; sex differences affect symptoms. |
| **[cp](ca://s?q=Explain_cp_in_heart_disease_dataset)** | Chest pain type (1–4) | Distinguishes angina vs. atypical pain; strong predictor. |
| **[trestbps](ca://s?q=Explain_trestbps_in_heart_disease_dataset)** | Resting blood pressure (mm Hg) | Hypertension is a major risk factor. |
| **[chol](ca://s?q=Explain_chol_in_heart_disease_dataset)** | Serum cholesterol (mg/dl) | High cholesterol accelerates atherosclerosis. |
| **[fbs](ca://s?q=Explain_fbs_in_heart_disease_dataset)** | Fasting blood sugar >120 mg/dl (1 = true, 0 = false) | Indicates diabetes risk, linked to heart disease. |
| **[restecg](ca://s?q=Explain_restecg_in_heart_disease_dataset)** | Resting ECG results (0–2) | Detects electrical abnormalities in heart function. |
| **[thalach](ca://s?q=Explain_thalach_in_heart_disease_dataset)** | Maximum heart rate achieved | Lower values may indicate poor cardiovascular fitness. |
| **[exang](ca://s?q=Explain_exang_in_heart_disease_dataset)** | Exercise-induced angina (1 = yes, 0 = no) | Symptom triggered by exertion, strong diagnostic clue. |
| **[oldpeak](ca://s?q=Explain_oldpeak_in_heart_disease_dataset)** | ST depression induced by exercise | Reflects ischemia (lack of blood flow to heart muscle). |
| **[slope](ca://s?q=Explain_slope_in_heart_disease_dataset)** | Slope of peak exercise ST segment (1–3) | Indicates exercise response; abnormal slopes suggest disease. |
| **[ca](ca://s?q=Explain_ca_in_heart_disease_dataset)** | Number of major vessels (0–3) colored by fluoroscopy | Direct measure of arterial blockage severity. |
| **[thal](ca://s?q=Explain_thal_in_heart_disease_dataset)** | Thallium stress test result (3, 6, 7) | Detects perfusion defects in heart muscle. |



In [ ]:
# drop duplicates based on specific columns
df_unque_chestpain = df.drop_duplicates(subset=['cp'])
print(df_unque_chestpain.shape)
df_unque_chestpain

(4, 14)


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
12,34,0,1,118,210,0,1,192,0,0.7,2,0,2,1
16,51,0,2,140,308,0,0,142,0,1.5,2,1,2,1
36,51,1,3,125,213,0,0,125,1,1.4,2,1,2,1


In [ ]:
df1 = pd.read_csv("/content/city_day.csv")
df1

,City,Date,PM2.5,PM10,NO,NO2,NOx,NH3,CO,SO2,O3,Benzene,Toluene,Xylene,AQI,AQI_Bucket
0,Ahmedabad,2015-01-01,NaN,NaN,0.92,18.22,17.15,NaN,0.92,27.64,133.36,0.00,0.02,0.00,NaN,NaN
1,Ahmedabad,2015-01-02,NaN,NaN,0.97,15.69,16.46,NaN,0.97,24.55,34.06,3.68,5.50,3.77,NaN,NaN
2,Ahmedabad,2015-01-03,NaN,NaN,17.40,19.30,29.70,NaN,17.40,29.07,30.70,6.80,16.40,2.25,NaN,NaN
3,Ahmedabad,2015-01-04,NaN,NaN,1.70,18.48,17.97,NaN,1.70,18.59,36.08,4.43,10.14,1.00,NaN,NaN
4,Ahmedabad,2015-01-05,NaN,NaN,22.10,21.42,37.76,NaN,22.10,39.33,39.31,7.01,18.89,2.78,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
29526,Visakhapatnam,2020-06-27,15.02,50.94,7.68,25.06,19.54,12.47,0.47,8.55,23.30,2.24,12.07,0.73,41.0,Good
29527,Visakhapatnam,2020-06-28,24.38,74.09,3.42,26.06,16.53,11.99,0.52,12.72,30.14,0.74,2.21,0.38,70.0,Satisfactory
29528,Visakhapatnam,2020-06-29,22.91,65.73,3.45,29.53,18.33,10.71,0.48,8.42,30.96,0.01,0.01,0.00,68.0,Satisfactory
29529,Visakhapatnam,2020-06-30,16.64,49.97,4.05,29.26,18.80,10.03,0.52,9.84,28.30,0.00,0.00,0.00,54.0,Satisfactory


In [ ]:
# Data Integration combine two datasets,  for that we need common field
df['ID'] = np.arange(len(df))
df1['ID'] = np.arange(len(df1))

In [ ]:
# Inner join on ID
combined = pd.merge(df, df1, on='ID', how='inner')
combined.shape

(1025, 31)

# Data Transformation
is process of changing the format, structure, or value of raw data to make clean, consistent, and ready for analysis or ml modles

In [ ]:
# clean hidden leading/ trailing spaces from column names
df.columns = df.columns.str.strip()


In [ ]:
#Automatically find the Age and Cholesterol columns regardless of casing or exact naming
age_col = next((col for col in df.columns if "age" in col.lower()), None)
chol_col = next((col for col in df.columns if "chol" in col.lower()),None)

In [ ]:
print(f"Age column: {age_col}")
print(f"Cholesterol column: {chol_col}")

Age column: age
Cholesterol column: chol


In [ ]:
# transformation age binning
if age_col:
  age_labels = ["Young", "Middle-aged", "Senior"]
  df['age_group'] = pd.cut(df[age_col], bins=[0, 30, 60, 100], labels=age_labels)
  # one-hot Encode the newly created age group
  df = pd.get_dummies(df, columns=['age_group'], dtype=int)
else:
  print("Warning: no age-related column found in dataset.")

In [ ]:
if chol_col:
  df["high_cholestrol"] = (df[chol_col] > 200).astype(int)
else:
  print("Warning: No cholsterol-related column found in dataset.")

In [ ]:
print("Tranformed data")
df

Tranformed data


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target,ID,age_group_Young,age_group_Middle-aged,age_group_Senior,high_cholestrol
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0,0,0,1,0,1
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0,1,0,1,0,1
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0,2,0,0,1,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0,3,0,0,1,1
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0,4,0,0,1,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1020,59,1,1,140,221,0,1,164,1,0.0,2,0,2,1,1020,0,1,0,1
1021,60,1,0,125,258,0,0,141,1,2.8,1,1,3,0,1021,0,1,0,1
1022,47,1,0,110,275,0,0,118,1,1.0,1,1,2,0,1022,0,1,0,1
1023,50,0,0,110,254,0,0,159,0,0.0,2,0,2,1,1023,0,1,0,1
